### PR TITLE
Compare the gradient consistency between GPU and CPU calculations

### LINKED ISSUE #3478: Need to compare the gradient consistency between GPU and CPU calculations.
Now the unit test for backward operator only performs the gradient check. But I think it can't guarantee the consistency between CPU and GPU calculations for the backward operator.

### PR DESCRIPTION
**Pull request by qingqing01**
Fix #3478 

1.  Compare the gradient consistency between GPU and CPU calculations in `gradient_checker.py`.
     And refine `gradient_checker.py`.
2. Move the unit test in `gradient_checker.py` into `test_gradient_checker.py`.
3. Add unit test for sigmoid backward implementation.

### CODE REVIEW

File: python/paddle/v2/framework/tests/gradient_checker.py [hunk 1]
```diff
@@ -109,12 +110,72 @@ def product(dim):
 
 
 class GradientChecker(unittest.TestCase):
-    def assert_is_close(self, numeric_grads, scope, max_relative_error,
-                        msg_prefix):
-        for name in numeric_grads:
-            b = numpy.array(scope.find_var(grad_var_name(name)).get_tensor())
-            a = numeric_grads[name]
+    def get_grad(self, forward_op, backward_op, input_vars, grad_names, place):
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/gradient_checker.py]: rename to `__get_grad`?
private method should name with `__xxx`

File: python/paddle/v2/framework/tests/gradient_checker.py [hunk 2]
```diff
@@ -109,12 +110,72 @@ def product(dim):
...
+        scope = core.Scope()
+        ctx = core.DeviceContext.create(place)
 
+        inputs = forward_op.inputs()
+        in_names = [item for k in inputs for item in inputs[k]]
+        outputs = forward_op.outputs()
+        out_names = [item for k in outputs for item in outputs[k]]
+
+        # create input var and set value
+        for name, value in input_vars.iteritems():
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/gradient_checker.py]: These `input_vars` is numpy array but not paddle's Variable, is it better to name it `input_python_vars`

File: python/paddle/v2/framework/tests/gradient_checker.py [hunk 3]
```diff
@@ -109,12 +110,72 @@ def product(dim):
...
+        backward_op.infer_shape(scope)
+        backward_op.run(scope, ctx)
+
+        outs = [
+            numpy.array(scope.find_var(name).get_tensor())
+            for name in grad_names
+        ]
+        return outs
+
+    def compare_grad(self, forward_op, inputs):
```
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/gradient_checker.py]: same as above `__compare_gradient`

File: python/paddle/v2/framework/tests/gradient_checker.py (hunk 1 above)
Comment (jacquesqiao) [inline python/paddle/v2/framework/tests/gradient_checker.py]: add some comment to these methods

Comment (qingqing01) [inline python/paddle/v2/framework/tests/gradient_checker.py]: Done. use `__get_gradient`.

File: python/paddle/v2/framework/tests/gradient_checker.py (hunk 2 above)
Comment (qingqing01) [inline python/paddle/v2/framework/tests/gradient_checker.py]: rename `input_value`, `input_python_vars` is a little long.

File: python/paddle/v2/framework/tests/gradient_checker.py (hunk 3 above)
Comment (qingqing01) [inline python/paddle/v2/framework/tests/gradient_checker.py]: This is not a private method. The user can call this method to compare the gradients for the backward op.

File: python/paddle/v2/framework/tests/gradient_checker.py (hunk 1 above)
Comment (qingqing01) [inline python/paddle/v2/framework/tests/gradient_checker.py]: Done.

**Pull request merged**